Support multi-target framework intellisense

mrward opened the pull request:
Type system service now supports multi-target frameworks.
Path bar supports multi-target framework selection.

<img width="831" alt="MultiTargetFrameworkDropdown" src="https://user-images.githubusercontent.com/372361/59188419-7f025200-8b6f-11e9-9c10-7b11c2c017b6.png">

<img width="975" alt="MultiTargetIntellisense" src="https://user-images.githubusercontent.com/372361/59188421-80cc1580-8b6f-11e9-844f-b9ea5a52af87.png">

Framework execution targets shown in main toolbar.

<img width="267" alt="ExecutionTarget" src="https://user-images.githubusercontent.com/372361/60883223-8a659d80-a241-11e9-8c69-d92881c280b4.png">

<img width="301" alt="SelectFrameworkExecutionTarget" src="https://user-images.githubusercontent.com/372361/60883226-8c2f6100-a241-11e9-8eb8-494d4c8bd9c9.png">

Fixes VSTS #572311 - Multi-framework support in Roslyn workspace
Fixes VSTS #572342 - Support picking target framework for intellisense
Fixes VSTS #937105 Show (Multiple TargetFrameworks) in project options
Fixes VSTS #572309 - Execution target support for multi-framework projects
Fixes VSTS #935387 - System.NullReferenceException exception in MonoDevelop.Projects.DotNetProject.<OnGetReferencedAssemblies>d__XXX.MoveNext() - Transitive references null reference exception fixed since this code no longer existx.
Fixes VSTS #572307 - Ensure navigation/refactoring operations on multi-frameworks projects are coalesced
Fixes VSTS #843603 Files with build action 'None' are still used by Intellisense

 - [x] Cache framework specific configuration
     - Ensure the cache is refreshed when project options page changes config settings.
 - [x] OnExecute project extension should be framework aware
        - Do not want ASP.NET Core project extension checking for certificate status when .NET Framework is the execution target.
  - [x] Text editor path bar does not update after changing from a single framework to multiple.
    - [x] See type system errors after saving the edited .csproj - [Gist](https://gist.github.com/mrward/75870b499d8c111597f5b41cf1ad4b53) 
    - [x] Changing frameworks in .csproj does not always update the path bar. Sometimes see no owner projects, sometimes no workspace changed event is fired.
        - [x] After the project file reload the OpenDocumentReferences stored in the type system have an associated Project but its ParentSolution is null. This results in TryOpenDocumentInWorkspace doing nothing which presumably then breaks the editor path bar.

mrward committed:
[Ide] Support multi-target framework projects in type system

Each target framework results in a Roslyn Project being created
with the references and source files specific to that target
framework.

The project info cache generates a separate cache file for each
target framework.

Running MSBuild targets now support the configuration having a
target framework. This was to avoid having to add a new framework
parameter to several project system  methods. The implementation
involves a custom configuration selector - a possible improvement
here would be to have these framework aware configurations
available from the Project itself - the current custom selector
implementation seems a little clumsy.

Project references on the Roslyn Project are not currently
multi-target framework aware.

Fixes VSTS #572311 - Multi-framework support in Roslyn workspace

mrward committed:
[CSharpBinding] Support multi-target frameworks in editor path bar

Use Roslyn Project information when adding projects to the text editor's
path bar. This allows multi-target framework projects to be supported.

Currently an extra project is being added to the path bar which has no
associated target framework. Visual Studio on Windows just shows a path
bar item for each target framework

Fixes VSTS #572342 - Support picking target framework for intellisense

mrward committed:
[Ide] Fix incorrect preprocessor symbols on switching framework

Switching to a different target framework in the text editor would
not update the preprocessor symbols available. The problem was that
the framework specific values were not being used when creating the
Roslyn project. Now a framework specific project configuration
is obtained from the project so the correct compiler options and
parse options are used by Roslyn.

mrward committed:
[Core] Resolve project references using MSBuild

The DotNetProject's GetReferencedAssemblies and GetReferences will
now resolve project references using MSBuild. Also switched to using
the design time targets. Previously ResolveAssemblyReferences was used.
ResolveAssemblyReferencesDesignTime and ResolveProjectReferencesDesignTime
are now used.

Using MSBuild to resolve project references will result in extra
metadata being added about the target framework. This will be used
when resolving the Roslyn project with multi-target projects and
avoids the IDE having to implement the same logic.

mrward committed:
[Ide] Fix Roslyn project references in multi-target projects

Using the NearestTargetFramework metadata returned from MSBuild
the correct Roslyn project id can be found for a project reference
when building the Roslyn workspace. Previously the target framework
was not used so a new Roslyn project id was being used.

mrward committed:
[Ide] Refactor type system's metadata reference handler

Previously the code was making two calls to the DotNetProject, one
for assembly references and one for project references. Now this is
done with one call to DotNetProject.GetReferencedAssemblies and
just one iteration of the references instead of two.

mrward committed:
[Ide] Fix extra Roslyn project showing in text editor path bar

With two target frameworks in a multi-target project, after reloading
the solution the text editor would show three projects. One for
each target framework and one extra project, which had no associated
target framework. The problem was that on reloading the projects
after using the cache was ignoring the target frameworks the project
had. This resulted in a new Roslyn ProjectId being created for
a null target framework.

mrward committed:
[DotNetCore] Rely on MSBuild to determine transitve project references

Removed code from the SdkProjectExtension which determined transitive
project references. This information is available from MSBuild if
a NuGet restore has been run.

Modified the transitive project reference tests so that they
run a NuGet restore

The ThreeProjectReferencesAndReferenceOutputAssemblyIsFalse test was
failing since MSBuild does not include project references that
have ReferenceOutputAssembly set to false. The test was expecting
such a reference to be returned.

mrward committed:
[Ide] Fix incorrect Roslyn project id used for multi-target project ref

Save the framework associated with the project reference to the
project info cache. This then ensures the correct Roslyn project id
is used when the cache is reloaded when the projects are multi-target
framework projects. Otherwise a new project id would be created for a
null target framework.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemFrameworkConfigurationSelector.cs:
```diff
@@ -0,0 +1,55 @@
+				var copy = new DotNetProjectConfiguration (name, platform, framework);
+				copy.CopyFrom (itemConfiguration);
+
+				return copy;
```
That is weird. A selector is supposed to return one of the configurations in the provided target, not create a new one. Why is this required?

mrward replied:
Honestly this is a bit of a hack which I do not like.

The framework information is needed when running an MSBuild target so I was trying to get this to the DotNetProject when it runs MSBuild to get references etc without adding an extra framework parameter to each method. Currently the active selector in the workbench has no knowledge of the framework so the type system service is cheating here and creates another selector and sets the framework it wants.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemConfiguration.cs:
```diff
@@ -84,17 +91,28 @@ public static void ParseConfigurationId (string id, out string name, out string
 		}
 
+		public string Framework {
+			get { return framework; }
```
Isn't the concept of framework is specific to .net? in that case this should be part of DotNetProjectConfiguration.

mrward committed:
[Core] Fix facades being added to wrong multi-target framework

If .NET Framework is the first target framework listed then facade
assemblies would be added to all the other target frameworks. This
would cause text editor errors when .NET Standard was selected due to
extra facade assemblies being added to the Roslyn project. Now the
TargetFramework associated with the DotNetProjectConfiguration is
used when resolving facade assemblies.

mrward committed:
[Core] Move target framework selection to DotNetProjectConfiguration

Moved the target framework out of ItemConfiguration to the
DotNetProjectConfiguration since it is DotNetProject specific.

mrward committed:
[Ide] Disable target framework selection for multi-target projects

In Project Options - Build - General the Target Framework combo box
now shows '(Multiple Frameworks)' and does not allow a framework to
be selected.

Fixes VSTS #937105 Show (Multiple Frameworks) in project options

mrward committed:
[Ide] Fix failing workspace files cache test

A telemetry project extension in md-addins is now calling the project's
GetReferencedAssemblies method during OnEndLoad so the test assertion
that this method is not called fails. This was breaking the
TestWorkspaceFilesCache_MultiTargetProjectReference_UsedOnLoad test.

Changed the test so it does not check if the GetReferencedAssemblies
method is called. If the cache was not used by the test the test
would hang called since a custom project extension used by the test
blocks the GetReferencedAssemblies call which ensures the cache
is being used. So there is no need to check if GetReferencedAssemblies
is called which avoids the problem with the telemetry project
extension calling this method.

mrward committed:
[Core] Add multi-target framework execution target support

Added basic support for executing different target frameworks. The
target frameworks are listed in the main toolbar. Selecting a
framework and then running the project will run the executable for
that framework.

No support for projects that have configuration specific OutputType.
No run configuration support for multiple target frameworks.

Fixes VSTS #572309 - Execution target support for multi-framework
projects

mrward committed:
[Ide] Fix References folder displayed for SDK style project

If the SDK style project targeted .NET Framework then the References
folder was displayed in the Solution window. The problem was that
the References folder was only hidden for .NET Core and .NET Standard
projects. Now it is hidden for all SDK style projects.

mrward committed:
[DotNetCore] Fix unable to run .NET Core project as second framework

If a multi-target framework project had .NET Framework as the first
framework it was not possible to run a .NET Core project using the
execution target selector in the main toolbar. The problem was that the
Project's TargetFramework was being checked by the
DotNetCoreProjectExtension to see if the project extension was
supported. Now all target frameworks are checked. The target frameworks
for the project are evaluated in DotNetProject.OnInitialize so that
a project extension can check the Project.TargetFrameworkMonikers
to see if supports any of the frameworks.

mrward committed:
[AspNetCore] Support multiple execution targets

mrward committed:
[Core] Include original short name with target framework monikers

Refactored the code so that the DotNetProject's TargetFrameworkMonikers
is used by the type system and project model instead of getting the
short names as an enumerable list of strings. The original short name
is copied to the TargetFrameworkMoniker so the full name of the
framework is available as well as the original short name.

mrward committed:
[DotNetCore] Project extensions now framework aware on project execution

The DotNetCoreProjectExtension and AspNetCoreProjectExtension were
taking action for the project's main target framework instead of the
target framework being executed.

An SdkProjectExtension can indicate which target frameworks it
supports when OnExecute is called. Only those target frameworks that
are supported will result in a new overload of OnExecute to be called
with the framework information.

mrward committed:
[DotNetCore] Execution command creation now framework aware

SdkProjectExtension derived classes can now override an overload
of OnCreateExecutionCommand which is passed the framework. This
method will only be called if the project extension supports that
framework. This moves the logic out of the individual project
extensions into the SdkProjectExtension when a project has multiple
target frameworks.

mrward committed:
[Ide] Fix project ids not removed from type system workspace

Removing a multi-target project from a solution would not remove the
associated project ids from the type system's workspace. The problem
was that an attempt was made to find the project ids from the mapping
based passing a null framework so no matches were found. Now the
frameworks are ignored and all project ids associated with the
project are removed.

mrward committed:
[Ide] Ensure ProjectId returned for non-framework aware code

The ProjectDataMap would return null if there was no match for a
null framework. Now the first ProjectId is returned. Existing code
is not multi-target framework aware so this changes the behaviour
back to the previous behaviour.

mrward committed:
[Ide] Fix problems reloading multi-target project in type system

When a project file is edited the project is reloaded. The type
system service was reloading the project info for each framweork
and calling the Roslyn workspace's OnProjectReloaded. If the
target framework has changed then this would fail since a new
project id was generated. New project ids are now added, instead
of being reloaded, and unused old project ids are removed.

mrward committed:
[Ide.Tests] Fix type system solution not being unloaded

Some tests were not unloading the solution from the type system
service.

mrward committed:
[Ide] Fix open documents referencing disposed project

When a project file is edited in the text editor and saved, triggering
a reload, the open documents registered with the type system would
not update their owner. This would happen if the document was
referenced by more than one text editor extension - the owner changing
would not remove the open document when the DocumentRegistration
class was disposed since another reference kept it alive. This
resulted in the C# path bar not being updated for multi-target
framework projects when the frameworks were changed and the project
file saved.

To fix this the owner of the open document is updated if it is not
null and does not match the original owner. This allows the type
system's TryRegisterOpenDocument to update the Roslyn workspace and
trigger the workspace changed event used by the C# path bar.

mrward committed:
[Core] Cache framework specific configurations

These get used fairly frequently by the type system service and
without being cached resulted in the MSBuild properties for the
project being evaluated each time.

mrward committed:
[Core] Make project configuration's framework short name internal

Renamed the property from Framework to FrameworkShortName since there
is an existing TargetFramework property.

Make FrameworkShortName property internal. The TargetFramework
property is public and can be used instead.

mrward committed:
[DotNetCore] Test node builders

Refactored the dependency node tests so they test the node builders
themselves. Previously the data object classes were being tested.
This is in preparation for new tests for multi-target projects.

mrward committed:
[DotNetCore] Show multi-target framework nodes as parent nodes

Previously multi-target framework projects would show the framework
nodes under Dependencies - NuGet and Dependencies - SDK. Now the
framework nows appear directly under the Dependencies folder. Note
that currently the Assemblies and Projects folders do not support
multi-target frameworks and just show all assemblies and projects
referenced directly in the project ignoring any conditions.

mrward committed:
[Core] Remove duplicate framework short name property from config

A different feature branch added TargetFrameworkShortName to the
DotNetProjectConfiguration. Re-use the same property instead of
having a different property for use with multi-target projects.

Therzok commented on main/src/addins/CSharpBinding/MonoDevelop.CSharp/CSharpPathedDocumentExtension.cs:
```diff
@@ -614,5 +618,32 @@ void CancelUpdatePath ()
 			src = new CancellationTokenSource ();
 		}
+
+		class OwnerProjectInfo
```
```suggestion
		sealed class OwnerProjectInfo
```

Therzok replied:
Maybe, since we implemented Equals and GetHashCode, we should also:
```suggestion
		class OwnerProjectInfo : IEquatable<T>
```

Therzok commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.NodeBuilders/DependenciesNodeBuilder.cs:
```diff
@@ -62,22 +63,37 @@ public override bool HasChildNodes (ITreeBuilder builder, object dataObject)
+		void OnPackageDependenciesChanged (object sender, EventArgs e)
+		{
+			var cache = (PackageDependencyNodeCache)sender;
+			var project = cache.Project;
```
Nit: cache.Project not used elsewhere, seems like a verbose local.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:
```diff
@@ -1193,50 +1230,15 @@ protected override async Task OnClearCachedData ()
-		{
-			var property = new MSBuildPropertyEvaluated (MSBuildProject, name, value, value);
-			metadata.SetProperty (name, property);
+			await base.OnClearCachedData ();
```
Not sure this method needs to be async anymore.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -1571,16 +1577,23 @@ static string[] GetTargetFrameworks (MSBuildProject project)
+		internal protected IEnumerable<string> GetTargetFrameworks ()
+		{
+			var frameworks = GetTargetFrameworks (MSBuildProject);
+			return frameworks ?? Enumerable.Empty<string> ();
```
```suggestion
			return frameworks ?? Array.Empty<string> ();
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs:
```diff
@@ -266,86 +267,108 @@ internal protected override async Task OnReevaluateProject (ProgressMonitor moni
-			var traversedProjects = new HashSet<string> (StringComparer.OrdinalIgnoreCase);
-			traversedProjects.Add (Project.ItemId);
+			string include = MSBuildProjectService.ToMSBuildPath (Project.ItemDirectory, fileName);
+			var globItems = Project.MSBuildProject.FindGlobItemsIncludingFile (include).ToList ();
```
Maybe we can do:
```suggestion
			var globItems = Project.MSBuildProject.FindGlobItemsIncludingFile (include).SingleOrDefault ();
```

To avoid matching all the items.

mrward replied:
Not sure I can change this. There may not be any matches - but could use FirstOrDefault instead. If there is more than one match we do not know what to return so fallback to asking the project for its default build action. Although not sure I have seen any examples of multiple matches here.

Therzok replied:
Yeah, was just trying to avoid a new collection allocation, as it seems redundant.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs:
```diff
@@ -266,86 +267,108 @@ internal protected override async Task OnReevaluateProject (ProgressMonitor moni
-					}
-					continue;
-				}
+			var targets = new List<ExecutionTarget> ();
```
```suggestion
			var targets = new List<ExecutionTarget> (Project.TargetFrameworkMonikers.Length);
```

Therzok commented:
LGTM!

mrward committed:
[CSharpBinding] Refactor OwnerProjectInfo

Update code after review. Seal class and implement IEquatable.

mrward committed:
[DotNetCore] Refactor node builder

Remove local variable in node builder after code review. The local
variable is not needed in the method.

mrward committed:
[Core] Improvements following review

Avoid allocations by using a value tuple as the key for the
framework specific configurations.

DotNetProject's OnClearCachedData does not need to be async.

Specify list length when creating the execution target list since it
is known.

Therzok commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.Gui/DotNetCoreRuntimeOptionsPanelWidget.cs:
```diff
@@ -49,43 +49,49 @@ public DotNetCoreRuntimeOptionsPanelWidget (DotNetProject project)
+					var cmp = string.CompareOrdinal (x.Id.Identifier, y.Id.Identifier);
+					if (cmp != 0)
+						return cmp;
+					return string.CompareOrdinal (y.Id.Version, x.Id.Version);
```
I know this was already broken in previous versions and not a big cause of worry for now, but, for example, given the same identifier, a framework list with versions `8.0`, `9.0`, `10.0` would be sorted incorrectly, as:
```
9.0
8.0
10.0
```

Filed https://devdiv.visualstudio.com/DevDiv/_workitems/edit/950504

Therzok replied:
Marking as resolved, as I don't think it's reasonable to expect a fix in this PR.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs:
```diff
@@ -0,0 +1,46 @@
+			Id = framework.ShortName;
+			Name = framework.ShortName;
+			Framework = framework;
+			Image = "md-framework-dependency";
```
Should we have a constant for this in Stock?

mrward replied:
This icon is currently in the NuGet addin. I guess it needs to be moved to Core since it is being used there.

mrward replied:
Making this a stock icon does not help here. Stock Icons are in MonoDevelop.Ide not MonoDevelop.Core so I cannot use it from here.

Therzok replied:
Gotcha! Thanks!

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:
```diff
@@ -2087,6 +2088,131 @@ protected override async Task OnReevaluateProject (ProgressMonitor monitor)
+				frameworkSpecificConfigurations [(name, platform, framework)] = configuration;
+		}
+
+		object frameworkSpecificConfigurationsLock = new object ();
```
```suggestion
		readonly object frameworkSpecificConfigurationsLock = new object ();
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/RuntimeOptionsPanel.cs:
```diff
@@ -72,46 +72,52 @@ public RuntimeOptionsPanelWidget (DotNetProject project, IEnumerable<ItemConfigu
+				for (int i = 0; i < frameworks.Count; i++) {
+					var fx = frameworks[i];
+					if (project.TargetFramework.Id == fx.Id) {
+						string name = notInstalled? GettextCatalog.GetString ("{0} (Not installed)", fx.Name) : fx.Name;
```
```suggestion
						string name = notInstalled ? GettextCatalog.GetString ("{0} (Not installed)", fx.Name) : fx.Name;
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.MetadataReferenceHandler.cs:
```diff
@@ -65,57 +65,70 @@ ImmutableArray<MonoDevelopMetadataReference> CreateDefaultMetadataReferences ()
-				var data = new AddMetadataReferencesData {
-					Result = new List<MonoDevelopMetadataReference> (),
+				var data = new AddReferencesData {
+					References = new List<MonoDevelopMetadataReference> (),
```
Note to file a bug to use `ImmutableArray.Builder`

Therzok replied:
Nothing to gain here, did some benchmarking.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectDataMap.cs:
```diff
@@ -48,56 +48,89 @@ public ProjectDataMap (MonoDevelopWorkspace workspace)
 				lock (gate) {
-					return projectIdMap.TryGetValue (p, out ProjectId result) ? result : null;
+					if (projectIdMap.TryGetValue (p, out var frameworkMappings)) {
+						var map = frameworkMappings.FirstOrDefault (f => f.Framework == framework);
```
Don't want to prematurely optimize, but `GetId` is called consistently a lot throughout the IDE usage.

Can we get rid of the lambda capture and LINQ usage?

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectDataMap.cs:
```diff
@@ -164,6 +214,18 @@ internal ProjectId[] GetProjectIds ()
 				}
 			}
+
+			class FrameworkMap
```
Nit:
```suggestion
			sealed class FrameworkMap
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs:
```diff
@@ -175,29 +199,45 @@ VersionStamp GetVersionStamp (MonoDevelop.Projects.Project project)
+				if (config == null || string.IsNullOrEmpty (framework))
+					return config;
+
+				return await dotNetProject.GetConfigurationAsync (config.Name, config.Platform, framework);
```
```suggestion
				return await dotNetProject.GetConfigurationAsync (config.Name, config.Platform, framework).ConfigureAwait (false);
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs:
```diff
@@ -97,11 +97,27 @@ class SolutionData
+				var projectId = projectMap.GetOrCreateId (p, oldProject, framework);
 
-				var config = GetDotNetProjectConfiguration (p);
+				var config = await GetDotNetProjectConfiguration (p, framework);
```
```suggestion
				var config = await GetDotNetProjectConfiguration (p, framework).ConfigureAwait (false);
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs:
```diff
@@ -1497,24 +1498,26 @@ void OnProjectModified (object sender, MonoDevelop.Projects.SolutionItemModified
-									Runtime.RunInMainThread (() => IdeServices.TypeSystemService.UpdateRegisteredOpenDocuments ()).Ignore();
+						foreach (string framework in ProjectHandler.GetFrameworks (project)) {
+							var projectInfo = ProjectHandler.LoadProject (project, cts.Token, null, null, framework).ContinueWith (t => {
+								if (t.IsCanceled)
```
Modify the ContinueWith call to harden and simplify the code a bit:

Specify `TaskScheduler` (`Task.Default` from the looks of it) for deterministic scheduling of the continuation.

Specify `TaskContinuationOptions.NotOnCanceled` and remove this IsCanceled check.

mrward replied:
What should TaskScheduler be set to?

Therzok replied:
TaskScheduler.Default. The code seems to want the continuation to run on the threadpool.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs:
```diff
@@ -511,14 +511,31 @@ async void OnSolutionItemAdded (object sender, MonoDevelop.Projects.SolutionItem
-						ws.OnProjectAdded (projectInfo);
+
+					foreach (string framework in ws.GetFrameworks (project)) {
+						var projectInfo = await ws.LoadProject (project, CancellationToken.None, oldProject, framework);
```
Can we not do this work on parallel?

mrward replied:
Just wondering if there is a benefit to doing it in parallel. Each framework would be locked when accessing project information, such as getting reference assemblies for that framework.

Therzok replied:
Hmmm, the only thing I'm thinking about here is that if we have 10 projects targeting 2 frameworks, we now do twice the work serially.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/WorkspaceFilesCache.cs:
```diff
@@ -78,34 +78,57 @@ void LoadCache (Solution sol)
+			}
+		}
+
+		internal IEnumerable<string> GetFrameworks (Project p)
```
Wonder if there is a way to deduplicate this into an extension method without polluting the internal API.

mrward replied:
For now I have changed the WorkspaceFilesCache to use the method on the MonoDevelopWorkspace.ProjectSystemHandler.

Therzok commented:
Done with review pass 1. Looks really good!

mrward committed:
[Ide] Address review comments

Fixed some code duplication.
Make some fields readonly.
Specify TaskScheduler for ContinueWith.
Avoid use of LINQ.

mrward committed:
[NuGet] Move framework icon to Ide

The Framework icon is used as the image for an execution target if
it is a framework belonging to a multi-target framework project.

mrward committed:
[Ide] Fix C# files with BuildAction of None used for code completion

The type system service was only checking the file extension not
the build action.

Fixes VSTS #843603 - Files with build action 'None' are still used by
Intellisense

Therzok commented:
LGTM iteration 2

mrward commented:
Test failure is happening on master so not specific to this branch.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -2919,21 +2932,21 @@ ProjectConfiguration LoadConfiguration (ProgressMonitor monitor, ConfigData cgrp
 		}
 
-		MSBuildProjectInstance CreateProjectInstaceForConfiguration (string conf, string platform, bool onlyEvaluateProperties = true)
+		protected internal MSBuildProjectInstance CreateProjectInstanceForConfiguration (string conf, string platform, string framework = null, bool onlyEvaluateProperties = true)
```
Does it need to be protected? It could just be internal.

mrward replied:
Changed it to be internal not protected.

mrward committed:
[Core] Make part of Project API internal not protected

Project.CreateProjectInstanceForConfigurationAsync does not need to
be protected since it is only used with MonoDevelop.Core.

Therzok commented:
Took it for a spin, works properly from what I've tested! Nice work!

sgmunn commented:
UI test for project system is from Nuget changes and is fixed in master.
.Net core tests are from mono bug. 

Neither failings are related as far as we know.

sgmunn reviewed (approved):
works good
